### ISSUE TITLE
Cyclic import detection when importing inside a function?

### ISSUE THREAD

**Issue by Holt59** [microsoft/pyright#4853]
**Describe the bug**
Pyright detects import cycle that I don't think are when importing inside a function.

**To Reproduce**

The following code will generate an import cycle on `from ..b.ba import BA` in `ab.py`:

```
Cycle detected in import chain
  ...\test-pyright\src\pkg\a\ab.py
  ...\test-pyright\src\pkg\b\ba.py
```

Code structure:

```
pkg/
  __init__.py
  a/
    __init__.py
    aa.py
    ab.py
  b/
    __init__.py
    ba.py
```

With content:

```python
# aa.py
class AA: pass

# ab.py
from ..b.ba import BA
from .aa import AA

class AB(AA):
    def f(self):
        ba = BA()

# ba.py
from ..a.aa import AA

class BA:
    def f(self) -> AA:
        from ..a.ab import AB
        return AB()
```

**Expected behavior**
I expected no "cycle import" error since in `BA`, only `AA` is imported globally and `AA` does not depends on `BA`.

**Screenshots or Code**

See above.

**VS Code extension or command-line**

I am running Pyright in VS Code through the Pylance extension, I have Pylance v2023.3.30.

**Additional contexts**

I've tried to reduce this to a minimal example, in my actual code, I have a `from ..xx import XXA` that generates the cycle, but if I change this to `from ..xx.xxa import XXA`, the cycle disappear, which I find weird.

**Comment by erictraut** [comment]
This is by design. Pyright's import cycle detection is intended to detect architectural layering violations — places where you are creating cycles in the call graph. Even if you hide such an import in a function, it's still an architectural layering violation. 

If you don't like how pyright's import cycle detection works, you can disable it. It's a completely optional check.

**Comment by Holt59** [comment]
> This is by design. Pyright's import cycle detection is intended to detect architectural layering violations — places where you are creating cycles in the call graph. Even if you hide such an import in a function, it's still an architectural layering violation.
> 
> If you don't like how pyright's import cycle detection works, you can disable it. It's a completely optional check.

Can you point to some complete explanation as to what is an import cycle for pyright if that's not the same definition as Python? 

The fact that `from ..xx import XXA` generates a cyclic import but not `from ..xx.xxa import XXA` does not makes it very hard for me to understand this "architectural layering" concept.

**Comment by erictraut** [comment]
In a program that honors architectural layering, can you create a graph of all modules and their import dependencies, and there will be no cycles in that graph.

Python can handle certain cycles at runtime if you use tricks like deferred imports (e.g. by importing from within a function). These tricks prevent your program from crashing. Determining exactly what is safe and what is not at runtime is difficult for a static analyzer because it depends on runtime execution ordering.

Pyright instead analyzes for the more restrictive definition of import cycle detection. If you don't want to enforce that in your code, you can leave this check disabled. You may be able to find other linters that check for the less restrictive import cycle detection if that's what you're looking for.

**Comment by Holt59** [comment]
> In a program that honors architectural layering, can you create a graph of all modules and their import dependencies, and there will be no cycles in that graph.
> 

The issue for me here is what pyright consider a dependency. For me, if I do `from ..xx.xxa import XXA` from a module, that module depends on `..xx` and `..xx.xxa` (since Python will have to execute `..xx` to access `..xx.xxa`, as far as I understand), but that does not seem to be the case in pyright since switching from `from ..xx import XXA` to `from ..xx.xxa import XXA` remove the error.
